app/routes.py

```diff
@@ -1,2 +1,238 @@
+        db.session.commit()
+
+        task_response = {"task": task.create_dict()}
+        return jsonify(task_response), 200
+
+
+#
+# Start Goals Routes
+#
+
+
+@goals_bp.route("", methods=["GET"])
+def get_all_goals():
+    title_from_url = request.args.get("title")
+
+    goals = Goal.goal_arguments(title_from_url)
+    goals_response = []
+    goals_response = [goal.create_dict() for goal in goals]
+
+    if not goals_response:
+        goals = Goal.query.all()
+        goals_response = [goal.create_dict() for goal in goals]
+
+    return jsonify(goals_response)
+
+
+@goals_bp.route("", methods=["POST"])
+def new_goal():
+    request_body = request.get_json()
+    if "title" not in request_body:
+        error_dict = {"details": "Invalid data"}
+        return jsonify(error_dict), 400
+
+    new_goal = Goal(title=request_body["title"])
+
+    db.session.add(new_goal)
+    db.session.commit()
+
+    goal_response = {"goal": new_goal.create_dict()}
+    return jsonify(goal_response), 201
+
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+def get_individual_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+    if not goal:
+        return jsonify(None), 404
+
+    if request.method == "GET":
+        goal_response = {"goal": goal.create_dict()}
+        return jsonify(goal_response), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+def replace_individual_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+    if not goal:
+        return jsonify(None), 404
+    form_data = request.get_json()
+
+    goal.title = form_data["title"]
+
+    db.session.commit()
+
+    goal_response = {"goal": goal.create_dict()}
+    return jsonify(goal_response), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"])
+def delete_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+    if not goal:
+        return jsonify(None), 404
+    db.session.delete(goal)
+    db.session.commit()
+
+    delete_message = {
+        "details": f'Goal {goal.goal_id} "{goal.title}" successfully deleted'}
+
+    return jsonify(delete_message), 200
+
+
+@goals_bp.route("<goal_id>/tasks", methods=["GET"])
+def get_goal_with_tasks(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if not goal:
+        return jsonify(None), 404
+
+    # goal_response = []
+    goal_response = [task.create_dict() for task in goal.tasks]
+
+    response = {"id": goal.goal_id,
+                "title": goal.title,
+                "tasks": goal_response}
+    return jsonify(response), 200
+
+
+@goals_bp.route("<goal_id>/tasks", methods=["POST"])
+def create_goal_with_tasks(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if not goal:
+        return jsonify(None), 404
+
+    form_data = request.get_json()
+    task_ids = form_data["task_ids"]
+    for id in task_ids:
+        task = Task.query.get(id)
+        if not task:
+            continue
+        task.goal = goal
+        db.session.commit()
+    
+    goal_response = [item.task_id for item in goal.tasks]
```

Comment:
Re-reading the tasks from the goal is a great decision, since we might have added tasks to a goal that already had tasks.